src/clients/UBAClient/UBAClientUtilities.ts

```diff
@@ -77,3 +83,296 @@ export async function getUBAFeeConfig(
+          relayerFee: {
+            relayerBalancingFee,
+            relayerCapitalFee: toBN(capitalFeeTotal),
+            relayerFee: toBN(relayFeeTotal).add(relayerBalancingFee),
+            relayerGasFee: toBN(gasFeeTotal),
+            amountTooLow: isAmountTooLow,
+          },
+          systemFee: {
+            depositBalancingFee,
+            lpFee,
+            systemFee: lpFee.add(depositBalancingFee),
+          },
+        });
+      }
+      // Verify that there is less than MAX_BUNDLE_CACHE_SIZE bundles in the cache
+      // If there are more, remove the oldest bundle
+      if (availableBundles.length > MAX_BUNDLE_CACHE_SIZE) {
+        availableBundles.shift();
+      }
+      // Set the bundles for this token
+      chainState.bundles[tokenSymbol] = availableBundles;
+    }
+    ubaChainStates[chainId] = chainState;
+  }
+  // Return the updated chain states
+  return ubaChainStates;
+}
+
+function getOpeningTokenBalances(
+  chainId: number,
+  spokePoolTokenAddress: string,
+  hubPoolClient: HubPoolClient,
+  hubPoolBlockNumber?: number
+): { blockNumber: number; spokePoolBalance: BigNumber; incentiveBalance: BigNumber } {
+  if (!isDefined(hubPoolBlockNumber)) {
+    if (!isDefined(hubPoolClient.latestBlockNumber)) {
+      throw new Error("Could not resolve latest block number for hub pool client");
+    }
+    hubPoolBlockNumber = hubPoolClient.latestBlockNumber;
+  }
+  const hubPoolToken = hubPoolClient.getL1TokenCounterpartAtBlock(chainId, spokePoolTokenAddress, hubPoolBlockNumber);
+  if (!isDefined(hubPoolToken)) {
+    throw new Error(`Could not resolve ${chainId} token ${spokePoolTokenAddress} at block ${hubPoolBlockNumber}`);
+  }
+  const balances = hubPoolClient.getRunningBalanceBeforeBlockForChain(hubPoolBlockNumber, chainId, hubPoolToken);
+  const endBlock = hubPoolClient.getLatestBundleEndBlockForChain([chainId], hubPoolBlockNumber, chainId);
+  return {
+    blockNumber: endBlock,
+    spokePoolBalance: balances.runningBalance,
+    incentiveBalance: balances.incentiveBalance,
+  };
+}
+
+function getFlows(
+  chainId: number,
+  chainIdIndices: number[],
+  spokePoolClients: { [chainId: number]: SpokePoolClient },
+  hubPoolClient: HubPoolClient,
+  fromBlock?: number,
+  toBlock?: number,
+  logger?: Logger
+): UbaFlow[] {
+  const spokePoolClient = spokePoolClients[chainId];
+
+  fromBlock = fromBlock ?? spokePoolClient.deploymentBlock;
+  toBlock = toBlock ?? spokePoolClient.latestBlockNumber;
+
+  // @todo: Fix these type assertions.
+  const deposits: UbaFlow[] = spokePoolClient
+    .getDeposits()
+    .filter(
+      (deposit: DepositWithBlock) =>
+        deposit.blockNumber >= (fromBlock as number) && deposit.blockNumber <= (toBlock as number)
+    );
+
+  // Filter out:
+  // - Fills that request refunds on a different chain.
+  // - Subsequent fills after an initial partial fill.
+  // - Slow fills.
+  const fills: UbaFlow[] = spokePoolClient.getFills().filter((fill: FillWithBlock) => {
```

Comment:
nit: will need to validate these fills used the correct `realizedLpFeePct` and also match perfectly with a deposit. See [here](https://github.com/across-protocol/sdk-v2/blob/master/src/utils/FillUtils.ts#L30) how we currently validate deposits. This assumes that the `SpokePoolClient` has mapped a `realizedLpFeePct` to each `DepositWithBlock` in its `update` call. This `realizedLpFeePct` obviously needs to be computed differently in the UBA but the same idea applies for validating a fill: it must match a deposit and have set the expected `realizedLpFeePct`.